python_modules/dagster/dagster/_core/execution/plan/state.py

```diff
@@ -97,6 +120,7 @@ class KnownExecutionState(
             ("step_output_versions", Sequence[StepOutputVersionData]),
             ("ready_outputs", Set[StepOutputHandle]),
             ("parent_state", Optional[PastExecutionState]),
+            ("asset_provenance", Sequence[AssetProvenanceData]),
```

Comment:
Any particular reason not to use a `Mapping[AssetKey, LogicalVersionProvenance]` here? It would enforce one-provenance-per-asset-key, which seems desirable.

Reply:
LOL I went down this exact road. I don't use it because it requires _another_ custom serializer (`AssetKey` as dict key).

We need some sugar for serializers, it's absurd how much code you have to write to convert one filed that has a non-standard dict key.

Reply:
Oof, right. I ran into this same issue when working on backfills. Would be cool to upgrade our serialization machinery support this.